python/cli/data_platform_workflows_cli/update_bundle.py

```diff
@@ -66,9 +76,16 @@ def main():
     # Full list of possible series config (unsupported) can be found under "Charm series" at https://juju.is/docs/olm/bundle
     default_series = file_data.get("series")
     for app in file_data["applications"].values():
-        app["revision"] = fetch_latest_revision(
-            app["charm"], app["channel"], app.get("series", default_series)
-        )
+        channel_map, metadata = fetch_charm_info_from_store(app['charm'], app['channel'])
+
+        if latest_revision := fetch_latest_revision(channel_map, app.get("series", default_series)):
+            app["revision"] = latest_revision
+        else:
+            raise ValueError(
+                f"Revision not found for {app['charm']} on {app['channel']} for Ubuntu {app.get('series', default_series)}"
+            )
+        if oci_image := fetch_oci_image_from_metadata(metadata):
+            app["resources"] = oci_image
```

Comment:
Now, if we want to facilitate things for the user a little bit (not much really, considering the overall amount of setup required) we could fetch the "OCI image metadata blobs" from each charm inside the bundle (its possible to do so using the snapcraft io API) and put them in a separate file inside the bundle repo (keeping them updated according to charm revisions)

This way, users wouldn't need to unpack each `.tar.gz` file to see which remote OCI images they must fetch and export using skopeo. WDYT?

Reply:
> does it contain the correct hash even if the charm is not using the latest resource revision in a channel? (e.g. we're using a charm revision that is not latest in channel?

Still not sure - in my local tests, the revision specified for each charm is being ignored and `store-admin` is exporting the latest rev in the channel by default. I'm in contact with the Store team to verify this + my understanding in the comment from before - keep you updated.

> IMO downloading the oci images should happen in the proxy or related tooling. I don't think we should upload oci images to the bundle git repos.

The download will still be done in the user side, my suggestion was to include the oci image's metadata in a separate file - ie a list of objects like this:
```
{
  "ImageName": "registry.jujucharms.com/charm/kotcfrohea62xreenq1q75n1lyspke0qkurhk/postgresql-image@sha256:8a72e1152d4a01cd9f469...",
  "Password": "MDAxOGxvY2F0aW9uIGNoYXJtc3Rvcm...",
  "Username": "docker-registry"
}
```

So that users know what images to download.

Reply:
UPDATE:

- As of right now, charmhub proxy doesn't have revision pinning of charms, see https://warthogs.atlassian.net/browse/PF-5185 - that's why my local testing wasn't working 🥲
- It also doesn't have a way to pin/specify resources. But they are working on a proposal, they may adhere to the syntax juju uses to pin revisions of resources: https://juju.is/docs/sdk/bundle.yaml#heading--applications-application-resources-resource

Reply:
> so is there currently no use for the update_bundle workflow? 

I mean, juju itself does consider charm revisions and resource pinnings when deploying bundles - It's the proxy store tooling that doesn't. But they do have plans to support it in the future, both [charm revisions](https://warthogs.atlassian.net/browse/PF-5185) and [resource revisions](https://warthogs.atlassian.net/browse/PF-5369)

Reply:
Long story here! Put a little TLDR in the PR description - basically the store doesn't support any sort of revision pinning as of now, but they will in the future, so we will do the pinnings as preparation on our part.